### PR TITLE
Code generation with ProxImaL

### PR DESCRIPTION
**Pull request by antonysigma**
Refactor proximal.algorithm.linearized_admm from Python to Halide language. The C++17 template function `LinearizedADMM::iterate()` describes the pipeline representing one iteration of the L-ADMM algorithm. Similarly,  `LinearizedADMM::computeConvergence()` for computing primal and dual residuals.

It is expected that the end-users modify the file `proximal/halide/src/user-problem/problem-config.h` containing the problem input and output data size. Also, implement the following pipelines:
* K.forward
* K.adjoint
* Omega_fns
* Psi_fns

in the file `problem-definition.h` . A Jinja2/Python template is provided at `proximal/templates` to facilitate the code generation in ProxImaL (code gen feature not yet implemented).

The Halide cross-compilation performs symbolic simplification to substitute all intermediate functions (e.g. K, omega_fns, psi_fns) into the ADMM algorithm:
image
and parallelize/vectorize intermedate stages where it makes sense.

The performance boost compared to existing ProxImaL varies upon the problem families; highest performance boost can be obtained if all absorbed proximal parameters (e.g. alpha, beta, omega, etc), as well as the CopyNode in `K`:
image
are completely eliminated by Halide. 

Note: perhaps someone should write a research article about it.

Resolves https://github.com/comp-imaging/ProxImaL/issues/57

### CODE REVIEW

File: proximal/halide/src/user-problem/test.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                  << '\n';
+    }
+
+    std::cout << "Top-left pixel = " << denoised(0, 0, 0) << '\n';
+
+    Buffer<float> output = std::move(denoised);
+    Halide::Tools::convert_and_save_image(output, "denoised.png");
+
+    return 0;
+}
```
Comment (antonysigma) [inline proximal/halide/src/user-problem/test.cpp:54]: @soufianekhiat This is the entry function to the baremetal algorithm, generated from `linearized-admm-gen.cpp` in the same folder. In this mock example, the Proximal Codegen generates a TV-regularized image denoising algorithm to run on CPU. See if it makes sense to you.

cc'ed @SteveDiamond .

File: proximal/halide/src/user-problem/problem-definition.h [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    Func adjoint(const FuncTuple<N>& u) {
+        /* Begin code-generation */
+        const Func z0 = KT_grad_mat(u[0], width, height);
+        const auto& z1 = u[1];
+        Func s;
+        s(x, y, c) = z0(x, y, c) + z1(x, y, c);
+        return s;
+        /* End code-generation */
```
Comment (antonysigma) [inline proximal/halide/src/user-problem/problem-definition.h:48]: The forward and backward mapping `K`. Refer to issue #59 for the challenges of exporting C++ code from Proximal.

**Review by SteveDiamond — APPROVED** [review]
This is amazing! It's a little beyond my ability to evaluate, but it looks good overall.

**Comment by jrk** [comment]
@antonysigma This is awesome! At the start of the project we aspired to one day get to end-to-end codegen, but never did, so this is really cool to see.

**Comment by antonysigma** [comment]
Thanks @jrk and @SteveDiamond for the complement. I applaud your effort to open source Halide and Proximal in pursuit of reproducible research.

The industry needs pioneers of 4th/5th generation language to solve complex problems. Without tools like these, ISP engineers and researcher out there are still stuck in polishing Matlab and LabVIEW in house.

Although this PR is a halfway solution to the end-to-end codegen, I believe by contributing back, someone else can eventually finish it.

**Pull request merged**